core/trino-main/src/main/java/io/trino/operator/TopNPeerGroupLookup.java

```diff
@@ -0,0 +1,374 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import io.trino.array.LongBigArray;
+import org.openjdk.jol.info.ClassLayout;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static it.unimi.dsi.fastutil.HashCommon.bigArraySize;
+import static it.unimi.dsi.fastutil.HashCommon.maxFill;
+import static it.unimi.dsi.fastutil.HashCommon.mix;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Optimized hash table for streaming Top N peer group lookup operations.
+ */
+// Note: this code was forked from fastutil (http://fastutil.di.unimi.it/) Long2LongOpenCustomHashMap.
+// Copyright (C) 2002-2019 Sebastiano Vigna
+public class TopNPeerGroupLookup
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(TopNPeerGroupLookup.class).instanceSize();
+
+    /**
+     * The buffer containing key and value data.
+     */
+    private Buffer buffer;
+
+    /**
+     * The mask for wrapping a position counter.
+     */
+    private long mask;
+
+    /**
+     * The hash strategy.
+     */
+    private final RowIdHashStrategy strategy;
+
+    /**
+     * The current allocated table size.
+     */
+    private long tableSize;
+
+    /**
+     * Threshold after which we rehash.
+     */
+    private long maxFill;
+
+    /**
+     * Number of entries in the set.
+     */
+    private long entryCount;
+
+    /**
```

Comment:
keep `fillFactor` below `maxFill`

Reply:
Sorry, what do you mean by this comment? It looks like fillFactor does come after maxFill in terms of line position.

Reply:
I mean directly below, e.g:
```
    /**
     * Threshold after which we rehash.
     */
    private long maxFill;
    /**
     * The acceptable load factor.
     */
    private final float fillFactor;
```